docs/design/2022-03-25-non-transactional-statement.md

```diff
@@ -0,0 +1,148 @@
+* [Motivation or Background](#motivation-or-background)
+* [Detailed Design](#detailed-design)
+* [Test Design](#test-design)
+    * [Functional Tests](#functional-tests)
+    * [Scenario Tests](#scenario-tests)
+    * [Compatibility Tests](#compatibility-tests)
+    * [Benchmark Tests](#benchmark-tests)
+* [Impacts & Risks](#impacts--risks)
+* [Investigation & Alternatives](#investigation--alternatives)
+* [Unresolved Questions](#unresolved-questions)
+
+## Introduction
+
+A new syntax that transforms a given DML into a serial execution of mutually exclusive collectively exhaustive statements. 
+
+For example, `split on a limit 1000 delete from t` can be transformed into `delete from t where a between 1 and 1000` and `delete from t where a between 1001 and 2000`, assuming the min and max of `a` are respectively 1 and 2000.
+
+## Motivation or Background
+
+Users need to do a bulk delete or update using a single statement, while currently TiDB cannot satisfy the requirement because of the performance issue, transaction size limit, and compatibility issues with tools.
+
+A new syntax is proposed to work around the problem. A non-transactional DML contains a DML and information that are used to "split" the statement, thus it is equivalent to a sequence of DMLs which is not transactional since it does not provide atomicity and probably isolation.
+
+Different from the deprecated batch-DML, a non-transactional DML splits SQLs so every split SQL is a normal statement and does not risk data-index consistency.
+
+## Detailed Design
+
+### User Interface
+
+#### Syntax
+The syntax: `split on <column_name> limit <batch_size> <DML>`
+
+In the first step only `delete` is going to be supported, but `update` and `insert into select` are also worth considering.
+
+The split column must be indexed. There are no other constraints on the DML.
+
+There can be dry run syntax to show the actual SQLs that will be executed. Query plans are not returned since there is no elegant way to contain both a SQL and a plan in a result set. 
+- `split on <column_name> limit <batch_size> dry run query <DML>` outputs the `SELECT` statement that will be executed.
+- `split on <column_name> limit <batch_size> dry run <DML>` outputs how the statement will be split.
+
+#### Output
+
+Users get feedback from SQL return values, logs and process infos.
+
+If all split statement succeed, the non-transactional DML returns a summary: the number of split statements and a success message. 
+If any of the split statement fails, the non-transactional DML returns the details and error messages of all failed jobs. 
+If the non-transactional DML is aborted, e.g. by `kill tidb`, it returns the error of context cancellation, and output details of failed jobs in logs.
+
+The `info` field in process info describes not only the current SQL that is executing, but also the progress of all jobs.
+Logs, slow logs and statement summaries will also contain the progress in the SQL text. 
```

Comment:
Do we consider binlog how to record this `split ...` statement? If it can't process appropriately, the `dm` tool may make inconsistent data between source and destination node.

Reply:
As far as I know, TiDB writes binlog for prewrite and commit requests. Then it should be unaware of the split statement?

Reply:
> As far as I know, TiDB writes binlog for prewrite and commit requests. Then it should be unaware of the split statement?

Yes, the `binlog` would be restored to SQL statements by the drainer and they could be like the dry run results. Anyway, it's better to sync with the data tool maintainers if the newly created grammar and execution behaviours would introduce any risks.